pkg/driver/vol_statter.go

```diff
@@ -0,0 +1,134 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package driver
+
+import (
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/volume/util/fs"
+	"time"
+)
+
+type volMetrics struct {
+	volPath string
+	timeStamp time.Time
+	volUsage []*csi.VolumeUsage
+}
+
+var (
+	volUsageCache = make(map[string]*volMetrics)
+	volStatterJobTracker = make (map[string]bool)
+)
+
+type VolStatter interface {
+	computeVolumeMetrics(volId, volPath string, refreshRate float64) (*volMetrics, error)
+	retrieveFromCache(volId string) (*volMetrics, bool)
+	removeFromCache(volId string)
+}
+
+type VolStatterImpl struct {
+}
+
+func NewVolStatter() VolStatter {
+	return &VolStatterImpl{}
+}
+
+func (v VolStatterImpl) computeVolumeMetrics(volId, volPath string, refreshRate float64) (*volMetrics, error){
+	if value, ok := v.retrieveFromCache(volId); ok {
+		if time.Since(value.timeStamp).Minutes() > refreshRate {
+			// Time to refresh volume stats
+			v.launchVolStatsRoutine(volId, volPath)
+		}
+		return value, nil
+	} else {
+		klog.V(4).Infof("Did not find volume metrics in cache for vol ID: %v , vol path: %v. Computing now!", volId, volPath)
+	}
+
+	v.launchVolStatsRoutine(volId, volPath)
+
+	// Return nil as kubelet might timeout waiting for volume stats
+	klog.Warningf("Volume metrics computation is underway for Vol ID: %v and metrics are not available yet.", volId)
+	return &volMetrics{
+		volPath: volPath,
+		timeStamp: time.Now(),
+		volUsage: []*csi.VolumeUsage{
+			{
+				Unit: csi.VolumeUsage_UNKNOWN,
+			},
+		},
+	}, nil
+}
+
+func (v VolStatterImpl) retrieveFromCache(volId string) (*volMetrics, bool) {
+	if value, ok := volUsageCache[volId]; ok {
+		return value, true
+	} else {
+		return nil, false
+	}
+}
+
+func (v VolStatterImpl) removeFromCache(volId string) {
+	delete(volUsageCache, volId)
+}
+
+func (v VolStatterImpl) launchVolStatsRoutine(volId, volPath string) {
+	if _, ok := volStatterJobTracker[volId]; ok {
+		klog.V(5).Infof("Volume stats computation job is underway for volume Id : %v. Awaiting results", volId)
+	} else {
+		volStatterJobTracker[volId] = true
+		go v.computeDiskUsage(volId, volPath)
+	}
+}
+
+func (v VolStatterImpl) computeDiskUsage(volId, volPath string) {
+	klog.V(5).Infof("Compute Volume Metrics invoked for Vol ID: %v", volId)
+
+	used, err := fs.DiskUsage(volPath)
+
+	if err != nil {
+		klog.Errorf("Failed to compute volume usage on path %s: %v",volPath, err)
+		return
+	}
+
+	volUsed, ok := used.AsInt64()
+
```

Comment:
nitpick, would like to see whitespace/newlines tightened up like here. Doesn't need to be a newline between the call and error checking. gofmt may help: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_aws-efs-csi-driver/238/pull-aws-efs-csi-driver-verify/1293265088089690114#1:build-log.txt%3A278

Reply:
Sure, will update this